feat: Quote Wrapper Pattern

jmuzina opened the pull request:
## Done

- Implements Quote Wrapper pattern per [design](https://www.figma.com/design/Hll6ZRvHyTTGBZZSoRd8ei/Higher-order-components-(HOCs)?node-id=1974-5068&m=dev) and [spec](https://docs.google.com/document/d/1-TMWuLvxmOKxPb8gITCEna6RVg05eGHKsB0EdcJ1w6Y/edit?tab=t.0)
- Drive-bys:
  - Adds notes to all patterns docs pages about the patterns being a recent/in-flux addition

Fixes [WD-13881](https://warthogs.atlassian.net/browse/WD-13881)

## QA
- Review all quote wrapper examples
  - **Large**
    - [All elements](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large)
    - [Without signpost](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-signpost)
    - [Without citation](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-citation)
    - [Without header](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-header)
    - [Without signpost and without header](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-header-no-signpost)
    - [Without CTA or image](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-cta-image?theme=light)
    - [Without CTA, image, or citation](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-no-cta-image-or-citation?theme=light)
    - [Minimal](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/large-minimal)
  - **Medium**
    - [All elements](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/medium)
    - [Without signpost](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/medium-no-signpost)
    - [Without citation](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/medium-no-citation)
    - [Without header](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/medium-no-header)
  - **Small**
    - [All elements](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/small?theme=light)
    - [Without signpost](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/small-no-signpost)
    - [Without citation](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/small-no-citation)
    - [Without header](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/small-no-header)
- Review [combined quote wrapper example](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/combined?theme=light) and verify all above examples are rendered as expected.
- Review [Quote Wrapper docs](https://vanilla-framework-5275.demos.haus/docs/patterns/quote-wrapper)
- Review updated [4.18.0 release notes](https://vanilla-framework-5275.demos.haus/docs/whats-new)

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [ ] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix release (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [x] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).


## Screenshots
<img width="1015" alt="Screenshot 2024-09-27 at 2 13 58 PM" src="https://github.com/user-attachments/assets/88cb5d6e-76cf-4434-be7a-de51c21fe821">
<img width="766" alt="Screenshot 2024-09-27 at 2 14 07 PM" src="https://github.com/user-attachments/assets/380b6441-664e-45d9-ba82-c0c5fa266882">
<img width="542" alt="Screenshot 2024-09-27 at 2 14 23 PM" src="https://github.com/user-attachments/assets/47dc775f-db90-4b72-8728-8bb3ee35b6f5">




[WD-13881]: https://warthogs.atlassian.net/browse/WD-13881?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://vanilla-framework-5275.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/9236)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,137 @@
+      {%- if has_heading_cta %}
+        <!--Optional heading CTA -->
+        <div class="col-3 col-medium-3 col-start-large-10 col-start-medium-5">
+          {#- TODO need a borderless variant of CTA block otherwise there is a top border here -#}
+          <div class="p-cta-block" style="border: none">
+           {{ heading_cta_content }}
+          </div>
```
We need to decide to either merge and use borderless CTA (from #5278) or just pass in link text and href and render a simple link here, without caring about text wrapping

jmuzina replied:
Adjusted this to simply use a slot for passing in link content

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,138 @@
+  - image (optional): An image to be displayed below the quote
+#}
+{%- macro quote_wrapper(
+  heading_text="",
```
Good call - adjusted

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,138 @@
+#}
+{%- macro quote_wrapper(
+  heading_text="",
+  quote_heading_level=2,
```
I'm hesitant to abstract this that far, in case design at some point wants us to support quote with text size that is not `<h2>` or `<h4>`.

bartaz replied:
I would also consider abstracting it. Especially that we already seem to limit that just to "2, 4, 6", so it's not like we allow any possible heading level.

It seems to be a quote size option. The fact that it matches to one of heading sizes is a bit of implementation detail.

I guess again, the question is: how is this documented and presented from design perspective - in figma. Is this marked as a heading level in Figma, or as a size of "Quote pattern" in Figma? Because in the end, that's what matters: developers will be looking into Figma, and need to translate this into props of a pattern.

jmuzina replied:
These were presented as heading levels in Figma. It is not difficult to translate them into a `quote_size` param with options of `large`, `medium`, `small`. However, that would make it difficult for this to support any other heading level sizes later.

jmuzina replied:
Abstracted this into `large`/`medium`/`small`, design does not think we will need more sizes ([ref](https://chat.canonical.com/canonical/pl/9qi85kurijr59n9yqu7dni7y5h))

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,138 @@
+  heading_text="",
+  quote_heading_level=2,
+  quote_text="",
+  citation_name="",
```
I suppose it's possible that we could be quoting something said by multiple people or a larger organization, rather than a specific person. 

Maybe this could be `citation_source_name` for added clarity without discouraging people from citing people or larger organizations?

jmuzina replied:
Updated to `citation_source_name` and `citation_source_title`

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,138 @@
+      <blockquote class="p-pull-quote u-no-margin--top u-no-padding--left">
+        <h{{ quote_heading_level }} class="u-no-padding--top">
+          <i>
+            “{{ quote_text }}”
```
Updated to use unicode quotation mark entities

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,138 @@
+
+      {%- if has_heading_link %}
+        <!--Optional heading link -->
+        <p class="col-3 col-medium-3 col-start-large-10 col-start-medium-5">
```
Good catch, not sure how that one got by the PR tests. Fixed!

jmuzina commented on templates/docs/examples/patterns/quote-wrapper/h2-no-citation.html:
```diff
@@ -0,0 +1,35 @@
+) -%}
+
+{%- if slot == 'heading_link' -%}
+  <a href="#">All the success stories ›</a>
```
Adjusted chevrons to use HTML entity

jmuzina commented on templates/docs/patterns/quote-wrapper/index.md:
```diff
@@ -0,0 +1,78 @@
+| Header                    | Header text & CTA. Useful for adding a title/link that describes all of a series of quotes. |
+| Signpost image            | Small image to display at the left/top of the quote, such as a logo or a portrait.          |
+| Image                     | Image to associate with the quote                                                           |
+| Citation name             | Name of the citation's source                                                               |
```
Updated to `Source name|title`

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,138 @@
+  {%- macro _quote_body() -%}
+    <div class="p-section--shallow">
+      <blockquote class="p-pull-quote u-no-margin--top u-no-padding--left">
+        <h{{ quote_heading_level }} class="u-no-padding--top">
```
Updated to use `p-heading--` classes rather than directly manipulating the tag name

jmuzina commented:
> There appears to be something going on here with the width of the `hr`/divider - at some screen sizes, it doesn't stretch all the way

@pastelcyborg Good catch, I think I've fixed that by moving the highlighted rule to outside of the grid, so it covers the full width of the whole pattern

jmuzina commented:
Made two changes to align with parts of the design I missed:

- Citation source name and title are now split on medium
- There is now an HR above the quote and citation on large size.

Placing the rule there is a bit tricky - I'll leave a separate inline review for team discussion

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,167 @@
+    {% endif -%}
+    {#- Skip to fourth column if the signpost_image (which takes up first three columns) is missing -#}
+    <div class="col-9{% if not has_signpost_image %} col-start-large-4{% endif %}">
+      <div class="u-hide--medium u-hide--small">
+        <hr>
+        {#- TODO find a better way to space the quote and citation away from the rule -#}
+        <div class="u-sv1"></div>
+      </div>
```
@pastelcyborg regarding my [previous comment](https://github.com/canonical/vanilla-framework/pull/5275#issuecomment-2289130361), here's the strangeness around spacing text away from the rule. 

The text of the quote and citation source name are aligned vertically because they all have no top padding. Adding a rule above them as-is makes the rule far too close to the text, so I use vertical spacing utility to compensate. Is this the sort of thing that will require additional scss?

jmuzina replied:
I think I had some weird cache weirdness going on when I reported this, as previously `hr` was getting no bottom margin, but now it is spacing the quote and citation out correctly as long as I don't use `p-rule`. So this is tangentially related to #5241 , but is not blocked by it.

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,167 @@
+        <div class="row">
+          <div class="col-2">
+            <div class="p-section--shallow">
+              {#- TODO find a better way to line the image top up with the quote and citation -#}
+              <div class="u-hide--medium u-hide--small u-sv1"></div>
+              {{ signpost_image_content }}
+            </div>
```
Per [previous comment](https://github.com/canonical/vanilla-framework/pull/5275#issuecomment-2289130361), this `u-sv-1` is used to make the top of the image line up with the top of the quote and citation

jmuzina commented:
Made an update to add a muted rule between quote and citation on small/medium. 

This will be broken until merge of #5241 , as `p-rule--muted` has no bottom margin currently.

jmuzina commented:
Using `hr.is-muted` so this isn't blocked by the HR PR - ready for review

jmuzina commented:
Made an update to add `hr.is-highlighted` (a highlighted rule that has bottom margin) and use it in this pattern.

lyubomir-popov commented:
I'll continue the review tomorrow, but the top line should not go wider than the fixed width:
![image](https://github.com/user-attachments/assets/2efa5dd6-b75f-4056-9724-5cd4c29ddc54)

lyubomir-popov commented:
- we should not tamper with baseline nudges (by adding u-no-padding--top/u-no-margin--bottom). This misalignes h2 and parsagraph for example. The coment applies across the board, not just this pr:

![image](https://github.com/user-attachments/assets/645fbed4-c513-4599-9025-0af50f5727d6)

- I need to give you a version of therexroth logo that aligns properly (by inserting white space in the image itself)

- these two should be wrapped by a common parent, with p-section class. Individual quotes in the parent should be spaced using the shallow spacing that is already there

![image](https://github.com/user-attachments/assets/7edfe028-9451-4c60-a27f-035ed63598ca)

- all hrs other than the one above the h2 heading should havethe class "is muted" (we need to add this to the hr docs)
![image](https://github.com/user-attachments/assets/16df12bb-399a-4978-ba57-823df3b433a7)

- can we please insert these mobile classes to prevent the logo from becoming huge on small screens:
![image](https://github.com/user-attachments/assets/81508dc2-c162-4d7f-b5db-0e75ed39b0b6)

jmuzina commented:
@lyubomir-popov Adjusted

Should the image be col-small-3 perhaps? It gets very small on small now:

<img width="282" alt="Screenshot 2024-08-20 at 9 28 41 AM" src="https://github.com/user-attachments/assets/58d6fbf2-437e-4991-95c4-ce83197996d3">

jmuzina commented:
>     * we should not tamper with baseline nudges (by adding u-no-padding--top/u-no-margin--bottom). This misalignes h2 and parsagraph for example. The coment applies across the board, not just this pr:

I got rid of all instances of `u-no-padding--top` and `u-no-margin--bottom` except for the following:

1. `h5.u-no-margin--bottom` on the citation source's name - without this, there is too much space between the citation source name and title to match the design
2. `p.u-no-margin--bottom` on the citation source's title - without this, there is too much space between the source and the content beneath it on small. I could also use `span` (without margin collapse), but that places the text slightly beneath the baseline. Here's how it looks with `p.u-no-margin--bottom`
<img width="239" alt="Screenshot 2024-08-20 at 4 41 01 PM" src="https://github.com/user-attachments/assets/f34f0a74-9112-4338-bcb8-050012b95077">

>     * I need to give you a version of therexroth logo that aligns properly (by inserting white space in the image itself)
Thanks! For the time being I've added you as an assignee while I await that logo.

>     * these two should be wrapped by a common parent, with p-section class. Individual quotes in the parent should be spaced using the shallow spacing that is already there
Done!

>     * all hrs other than the one above the h2 heading should havethe class "is muted" (we need to add this to the hr docs)
Done (changed HR between quote and heading to `hr.is-muted`). 
Also, the [base HR docs](https://vanillaframework.io/docs/base/separators#muted-horizontal-rule) do say this:
"By default, all horizontal rules should be muted. The regular, non-muted variant is only used at the top of content sections, above headings", if that isn't enough let me know and we can address it.

>     * can we please insert these mobile classes to prevent the logo from becoming huge on small screens:
Done, but used `col-small-2` rather than `col-small-1` as discussed at standup today, to prevent the logo from being excessively small.

lyubomir-popov commented:
@Sophie-32 is there a quote we can use in the vanilla examples? we're currently using rexroth, we probably don't have permisison. Ideally, one of our own with a canonical product logo?

jmuzina commented:
@lyubomir-popov Requested an H6 variant, adding that shortly - this can be a very trivial addition

jmuzina commented:
Quote wrapper variant added. I've also added an example showing the minimal content (just the quote text).

jmuzina commented:
@lyubomir-popov @pastelcyborg @advl I've added notices to the docs pages of all patterns examples stating that patterns are a recent/changing addition:

<img width="916" alt="Screenshot 2024-08-21 at 10 47 45 AM" src="https://github.com/user-attachments/assets/ab45abc9-c4d9-4d99-8255-2827b50cdc35">

jmuzina commented on templates/docs/patterns/quote-wrapper/index.md:
```diff
@@ -0,0 +1,70 @@
+  title: Quote Wrapper | Patterns
+---
+
+<div class="p-notification--information">
```
These docs pages are rendered with [mistune](https://mistune.lepture.com/en/latest/guide.html), not Jinja - so we'd need to research & implement some new way to do this. So, it may be a bit out of scope of this PR, but I do agree this is cumbersome

jmuzina replied:
@pastelcyborg I found a way to do this, and I've implemented it. Let me know what you think.
It turns out this can be done with Jinja just like any other template.

jmuzina commented on templates/docs/patterns/hero/index.md:
```diff
@@ -4,6 +4,13 @@ context:
+<div class="p-notification--information">
+  <div class="p-notification__content">
+    <h3 class="p-notification__title">Work in progress:</h3>
+    <p class="p-notification__message">Vanilla's pattern examples are newly released and still evolving as we receive feedback.</p>
```
Good call, reworded

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,174 @@
+      <blockquote class="p-pull-quote u-no-margin--top u-no-padding--left">
+        <p class="p-heading--{{ quote_heading_level }}">
+          <i>
+            &#8220{{ quote_text }}&#8221
```
Good call, added semicolons there

lyubomir-popov commented:
I think I found a suitable quote to replace the rexroth content - [here ](https://canonical.com/careers/company-culture/diversity)
![image](https://github.com/user-attachments/assets/42ab9627-f9c3-4fbf-b7b6-8b2b9a6048ab)

jmuzina commented:
@lyubomir-popov I've updated the following:

- Revised the H6 quote to a quote from the 20 years page, provided by @mattea-turic
- Replaced the Rexroth signpost logos with Qualcomm logos (from the 20 years page)
- Replaced the Rexroth full-width cover images with the full-width cover image from our [diversity](https://canonical.com/careers/company-culture/diversity) page

lyubomir-popov commented:
this overrides the h4 margin, which is incorrect - can we remove pls 

![image](https://github.com/user-attachments/assets/b377eb43-3616-4008-bf61-1c19fd3ae727)

same on the other text 
![image](https://github.com/user-attachments/assets/28e505ba-140a-403b-840e-7fc91a5e28c7)

and we need to use canonical quotes from the page I sent you yesterday, as now the logo, person and role don't match

all of this is also old styling
![image](https://github.com/user-attachments/assets/166626ef-19c0-4635-ad56-26dea0ae93de)

lyubomir-popov commented:
updated logos to match text baseline at the 3 font-sizes https://assets.ubuntu.com/manager?tag=quote-example-logos

jmuzina commented:
@lyubomir-popov I've updated the macro to remove the spacing issues that you raised and we discussed in standup, and changed the quote text and citation sources to Canonical personnel from the careers pages.

I've also changed the CTA to "Explore Careers" and the header to "Our hiring process".

Have a look at the [combined example](https://vanilla-framework-5275.demos.haus/docs/examples/patterns/quote-wrapper/combined?theme=light) and verify every example looks correct, please :)

lyubomir-popov commented:
no need for u-no-margin--bottom on these 

![image](https://github.com/user-attachments/assets/ba0524cf-4170-40f6-a5a5-a088a35a531a)

once removed, the baselines of the h4 and second line of the role line up, which is an added nice touch

![image](https://github.com/user-attachments/assets/825ee76c-b15d-4dac-b5c7-7aa54d88964c)

same here:

![image](https://github.com/user-attachments/assets/f711889e-9e49-4582-bea6-7cd91618c074)

Do we have an example with the header with black line not present when the quote is other than the first?

jmuzina commented:
> no need for u-no-margin--bottom on these
> 
> ![image](https://private-user-images.githubusercontent.com/2741678/360499702-ba0524cf-4170-40f6-a5a5-a088a35a531a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQzMzgyNjMsIm5iZiI6MTcyNDMzNzk2MywicGF0aCI6Ii8yNzQxNjc4LzM2MDQ5OTcwMi1iYTA1MjRjZi00MTcwLTQwZjYtYTVhNS1hMDg4YTM1YTUzMWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjJUMTQ0NjAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWQ3OWIxNmNkMzI5ZWM5NzljZGE1ZTI1NGU4ZmU1NjY0MmZlNmM4NGQyNGIyMDcyNDk4OGVlNWM0ZWY3MDcxZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.SywKOq8ZyydLXi5gdKogWbX0TktZUBc5wBFduFEjWa4)
> 
> once removed, the baselines of the h4 and second line of the role line up, which is an added nice touch
> 
> ![image](https://private-user-images.githubusercontent.com/2741678/360500486-825ee76c-b15d-4dac-b5c7-7aa54d88964c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQzMzgyNjMsIm5iZiI6MTcyNDMzNzk2MywicGF0aCI6Ii8yNzQxNjc4LzM2MDUwMDQ4Ni04MjVlZTc2Yy1iMTVkLTRkYWMtYjVjNy03YWE1NGQ4ODk2NGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjJUMTQ0NjAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzFkYTM3ZGFjMDY2MDViODFlMTFiNzY3NTRhOTJkMmNmZDU2OGIzYWY5NmJjZWJjNTkyYTFjODk4MGUxOGVlYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.pjPeefH-kxLx4vNqSKAs4xRcMj9VPib3hpaOYYtNQ-U)
> 
> same here:
> 
> ![image](https://private-user-images.githubusercontent.com/2741678/360501319-f711889e-9e49-4582-bea6-7cd91618c074.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQzMzgyNjMsIm5iZiI6MTcyNDMzNzk2MywicGF0aCI6Ii8yNzQxNjc4LzM2MDUwMTMxOS1mNzExODg5ZS05ZTQ5LTQ1ODItYmVhNi03Y2Q5MTYxOGMwNzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjJUMTQ0NjAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2ZiZDc0NTA5YzA5YjczYTZkZTY1MjY3NDcxMWFjNzllYmVhOWYyNWEzZDBhNDM2OThjOWZiNWRlNWM0MzE5ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.1b0U1xk4weWzSdZA__8SJ5k4cSSM_YvM4HA700QseoA)
> 
> Do we have an example with the header with black line not present when the quote is other than the first?

Just added example of no-header for each of the heading levels

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,167 @@
+{%- macro quote_wrapper(
+  title="",
+  quote_heading_level=2,
+  quote_text="",
```
I think all params that accept text input that will be rendered as content should end in `_text` - it's a good way to suggest to the user what the intent is there.

jmuzina replied:
I've switched all of the quote wrapper's text-only parameters (so, everything but `quote_heading_level`) to end with `_text`. The other patterns should be handled separately, I think - this PR is already quite large. I'll make an issue to cover Hero & Tiered List

jmuzina commented on templates/docs/examples/patterns/quote-wrapper/h2-no-citation.html:
```diff
@@ -0,0 +1,35 @@
+  quote_text="This is a company where intelligence and learning are highly valued."
+) -%}
+
+{%- if slot == 'heading_link' -%}
```
I'm of two minds on this sort of convention for slots as they can be much more free-form. 

1. Consider a "description" block that is meant to contain free-form descriptive elements (mostly `<p>`). The information passed into that slot is not necessarily the description's "content", it *is* the description in and of itself. The information there is "content" in the sense of it being "slot content" - but that is more of a concern for the people making the macros than it is those using the macros.
2. In the literal sense of the word, the information passed into the slots is site content. 

I guess I'm just not sure if we can say for certain that every slot should be called `_content`.

jmuzina replied:
Here's an example of why I'm hesitant to do this:

<img width="354" alt="Screenshot 2024-08-22 at 3 40 49 PM" src="https://github.com/user-attachments/assets/1bb7ef69-e2db-440f-8741-853da9140783">

This slot is intended for just passing in a link. It feels a bit strange to say `link_content`, as the developer is passing in a link, not the contents of a link.

bartaz replied:
Just to add to that: if every slot name ends with `_content`, what's the point?

jmuzina commented:
@lyubomir-popov Is hoping to decrease spacing between quote source name and role slightly. Here are two proposed approaches:

Add `u-no-margin--bottom` to both of them, then wrap their row in a shallow section:
<img width="937" alt="Screenshot 2024-08-23 at 9 52 28 AM" src="https://github.com/user-attachments/assets/c3d152f8-55ff-4cbf-b6d8-f20b9e8176e4">

OR:

The following change to markup around the name and role:
<img width="323" alt="Screenshot 2024-08-23 at 10 01 12 AM" src="https://github.com/user-attachments/assets/85ded183-d017-4624-bf2f-b334fb6e4441">

jmuzina commented:
@lyubomir-popov I made the change requested in [Figma](https://www.figma.com/design/vH19y2J2YqwUfksebSUvTW?node-id=26-1135&m=dev#923221984) and tried to make it work with the spacing adjustments, have a look and let me know what you think.

`u-no-margin--bottom` was added to the source name to align with what is shown in Figma

bartaz commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,178 @@
+      <hr class="u-hide--large is-muted">
+      {%- if has_citation_source_name and (has_citation_source_title or has_citation_source_organisation) %}
+        <!-- When source name and title/org are both present, wrap them in a 50/50 row to split them on medium -->
+        <div class="row">
```
Why not `row--50-50`?

jmuzina commented on templates/docs/templates/notice.html:
```diff
@@ -0,0 +1,7 @@
+<!-- This template is used to generate component deprecation/information notices. -->
```
Good point, that can be changed to macro

jmuzina replied:
Changed to use macro

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,178 @@
+
+  <div class="p-section">
+    {{- _heading_block() -}}
+    <div class="row">
```
@bartaz  When I use `row--25-75`, the nested grid for the source is broken:

Expected:
![Screenshot from 2024-08-27 11-42-11](https://github.com/user-attachments/assets/d10ecffe-6867-4195-9cb9-da92644e3c0a)

Observed:
![Screenshot from 2024-08-27 11-41-56](https://github.com/user-attachments/assets/bfa73c4f-9d9a-4f40-9235-1af4a7ccb1cc)

jmuzina replied:
Update: Lyubo's made a [design simplification](https://www.figma.com/design/vH19y2J2YqwUfksebSUvTW?node-id=26-1135&m=dev#923221984) that means this can be done

jmuzina replied:
Switched to 25/75 row

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,178 @@
+          </div>
+        </div>
+      {% endif -%}
+      {#- Skip to fourth column if the signpost_image (which takes up first three columns) is missing -#}
```
See [above comment](https://github.com/canonical/vanilla-framework/pull/5275#discussion_r1733123428)

jmuzina replied:
Switched to 25/75 row

jmuzina commented on templates/_macros/quote-wrapper.jinja:
```diff
@@ -0,0 +1,172 @@
+          </div>
+        </div>
+      {% endif -%}
+      {#- Skip to fourth column if the signpost_image (which takes up first three columns) is missing -#}
```
Note to self tomorrow: this comment can be removed, as we're now using the 25/75 row, making this obsolete

jmuzina replied:
Updated

jmuzina commented on templates/_macros/quote-wrapper.jinja:
Need to rename this file and namespace the macro with vf_ for compatibility with macro publishing efforts

jmuzina replied:
Updated

bartaz commented on side-navigation.yaml:
```diff
@@ -117,6 +117,8 @@
       url: /docs/patterns/hero
     - title: Tiered list
       url: /docs/patterns/tiered-list
+    - title: Quote Wrapper
```
If "Tiered list" I guess it should be "Quote wrapper" then
```suggestion
    - title: Quote wrapper
```

bartaz commented on templates/_macros/hero.jinja:
```diff
@@ -72,10 +72,7 @@
-      {#- On 25-75 signpost, header top should align with top of the signpost to its left. -#}
-      {% set title_class = "u-no-padding--top" %}
-    {%- elif has_full_width_image -%}
+    {%- if has_full_width_image -%}
```
Not sure how this change is related to quote?

jmuzina replied:
This is a drive-by to fix something we missed in design review on hero, as requested by Lyubo [here](https://chat.canonical.com/canonical/pl/1bda8imr5bymbde47unn8hjmoe)

bartaz commented on templates/docs/patterns/quote-wrapper/index.md:
```diff
@@ -0,0 +1,68 @@
+| Source               | Name, title, and organisation of the quote's source                                         |
+
+All examples shown here use `<h2>`-sized quote text. However, this should be changed according to text length.
+For short quotes, use `.p-heading--2`.
```
Should we document heading sizes here, or macro size values?

jmuzina replied:
In the [PR that shows macro calls in code snippets](https://github.com/canonical/vanilla-framework/pull/5321), we show HTML by default still. So, I figure the best way to do this is to document the HTML classes in the MD sections, and document the Jinja choices using table of params/slots as with #5325

jmuzina replied:
Updated to use macro param/slot documentation per established standards from hero and tiered list

bartaz reviewed (commented):
Overall it looks good.

But in relation to recent discussion on the shape of these patterns (and how flexible they are, etc) I think it would be good to have a spec approved first as well.
Do we have the spec (shape of the parameters, variants, options, etc) written down? It would be good for it to be reviewed and approved by Lyubo, UX, ideally Sophie as well. To avoid a situation when we release this pattern and next thing we get a request to add a layout option or additional image/cta block or whatever.

jmuzina commented:
The team has decided in SU today that we will be starting our HOC speccing efforts by retroactively speccing our to-be-delivered HOCs before they are merged. So, this PR will be blocked until review and approval of a spec by design, UX, and code.

jmuzina commented:
Made some changes after meeting with @lyubomir-popov to discuss [c.com PR 1380](https://github.com/canonical/canonical.com/pull/1380).

- Design change: Changed the medium layout design: it now uses a 33/66 split with the signpost (or a gap if no signpost is provided) on the left, rest of the content on the right. The header's grid was also adjusted to use 33/66 on medium to line up with the signpost/quote grid.
- Wrapping the quote text and the signpost logo in a shallow section
- Slightly reduced spacing between the header row and rest of the quote
- Fixed some broken grid usage for the signpost logo
- Switched from using base `hr` to `p-rule` component
- Show the HR above the quote text on all breakpoints instead of only on large

jmuzina commented:
Spec has been approved, putting this PR back into consideration :)

bartaz commented:
I guess it could be worth updating the pattern to use new 8 col grid? But also, doesn't need to be in scope of this PR. We could address it as follow up (could be done for all patterns I suppose).

bartaz reviewed (approved):
Looks good. We could consider updating this (and other?) patterns to new grid, but this can be a follow up.
